File: ios/Solace/ViewModels/OnboardingViewModel.swift
import SwiftUI
import Combine

@MainActor
public final class OnboardingViewModel: ObservableObject {

    // Steps — remotely configurable; defaults baked in
    @Published public private(set) var steps: [OnboardingStep] = DefaultOnboardingSteps
    @Published public private(set) var currentIndex: Int = 0
    @Published public var answers: OnboardingAnswers = OnboardingAnswers()
    @Published public private(set) var isComplete: Bool = false
    @Published public private(set) var isSaving: Bool = false
    @Published public private(set) var error: String? = nil

    private let repo = OnboardingRepository.shared

    // MARK: - Derived state

    public var currentStep: OnboardingStep { steps[currentIndex] }
    public var progress: Double { Double(currentIndex + 1) / Double(steps.count) }
    public var isFirstStep: Bool { currentIndex == 0 }
    public var isLastStep: Bool { currentIndex == steps.count - 1 }

    public var canAdvance: Bool {
        switch currentStep.type {
        case .textInput:
            switch currentStep.id {
            case "nickname":         return answers.username.count >= 2
            case "feeling_sentence": return !answers.feelingSentence.isEmpty
            default:                 return true
            }
        case .multiSelectChips:
            guard currentStep.isRequired else { return true }
            return selectionsFor(step: currentStep).count >= currentStep.minSelections
        case .colorPicker: return true
        case .radioList:
            guard currentStep.isRequired else { return true }
            return !radioAnswerFor(step: currentStep).isEmpty
        case .milestone: return true
        }
    }

    // MARK: - Navigation

    public func advance() {
        guard currentIndex < steps.count - 1 else { return }
        currentIndex += 1
    }

    public func back() {
        guard currentIndex > 0 else { return }
        currentIndex -= 1
    }

    // MARK: - Answer setters

    public func setMoodColor(_ key: MoodColorKey) {
        answers.moodColor = key
        answers.moodLabel = defaultMoodPalette.first { $0.key == key }?.label ?? ""
    }

    public func toggleChip(stepId: String, option: String, max: Int) {
        switch stepId {
        case "mood_descriptor":  toggle(&answers.moodDescriptors, option, max: max)
        case "feeling_now":      toggle(&answers.feelingNow,      option, max: max)
        case "feeling_like":     toggle(&answers.feelingLike,     option, max: max)
        case "struggling_with":  toggle(&answers.strugglingWith,  option, max: max)
        case "vibes":            toggle(&answers.vibes,           option, max: max)
        case "hope_to_gain":     toggle(&answers.hopeToGain,      option, max: max)
        default: break
        }
    }

    public func setRadioAnswer(stepId: String, option: String) {
        switch stepId {
        case "sought_support": answers.soughtSupport = option
        case "openness":       answers.openness = option
        default: break
        }
    }

    // MARK: - Completion

    public func completeOnboarding(uid: String) {
        isSaving = true
        Task {
            do {
                try await repo.saveOnboardingAnswers(uid: uid, answers: answers)
                isComplete = true
            } catch {
                self.error = error.localizedDescription
            }
            isSaving = false
        }
    }

    public func clearError() { error = nil }

    // MARK: - Helpers

    public func selectionsFor(step: OnboardingStep) -> [String] {
        switch step.id {
        case "mood_descriptor":  return answers.moodDescriptors
        case "feeling_now":      return answers.feelingNow
        case "feeling_like":     return answers.feelingLike
        case "struggling_with":  return answers.strugglingWith
        case "vibes":            return answers.vibes
        case "hope_to_gain":     return answers.hopeToGain
        default:                 return []
        }
    }

    public func radioAnswerFor(step: OnboardingStep) -> String {
        switch step.id {
        case "sought_support": return answers.soughtSupport
        case "openness":       return answers.openness
        default:               return ""
        }
    }

    private func toggle(_ list: inout [String], _ item: String, max: Int) {
        if let i = list.firstIndex(of: item) {
            list.remove(at: i)
        } else if max == 0 || list.count < max {
            list.append(item)
        }
    }
}

File: android/app/src/main/java/com/solace/ui/onboarding/OnboardingViewModel.kt
package com.solace.ui.onboarding

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.solace.data.onboarding.OnboardingRepository
import com.solace.ui.config.DefaultMoodPalette
import com.solace.ui.config.MoodColorKey
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import javax.inject.Inject

data class OnboardingUiState(
    val steps: List<OnboardingStep> = DefaultOnboardingSteps,
    val currentIndex: Int = 0,
    val answers: OnboardingAnswers = OnboardingAnswers(),
    val isComplete: Boolean = false,
    val isSaving: Boolean = false,
    val error: String? = null,
) {
    val currentStep: OnboardingStep get() = steps[currentIndex]
    val progress: Float get() = (currentIndex + 1).toFloat() / steps.size.toFloat()
    val isFirstStep: Boolean get() = currentIndex == 0
    val isLastStep: Boolean get() = currentIndex == steps.lastIndex

    /** Whether the current step has a valid answer to advance. */
    val canAdvance: Boolean get() = when (currentStep.type) {
        OnboardingStepType.TEXT_INPUT -> when (currentStep.id) {
            "nickname" -> answers.username.length >= 2
            "feeling_sentence" -> answers.feelingSentence.isNotBlank()
            else -> true
        }
        OnboardingStepType.MULTI_SELECT_CHIPS -> {
            val selections = currentStep.currentSelections(answers)
            !currentStep.isRequired || selections.size >= currentStep.minSelections
        }
        OnboardingStepType.COLOR_PICKER -> true
        OnboardingStepType.RADIO_LIST -> {
            val selection = currentStep.currentRadioAnswer(answers)
            !currentStep.isRequired || selection.isNotBlank()
        }
        OnboardingStepType.MILESTONE -> true
    }
}

// Extension helpers to read answers per step
private fun OnboardingStep.currentSelections(answers: OnboardingAnswers): List<String> = when (id) {
    "mood_descriptor"  -> answers.moodDescriptors
    "feeling_now"      -> answers.feelingNow
    "feeling_like"     -> answers.feelingLike
    "struggling_with"  -> answers.strugglingWith
    "vibes"            -> answers.vibes
    "hope_to_gain"     -> answers.hopeToGain
    else               -> emptyList()
}

private fun OnboardingStep.currentRadioAnswer(answers: OnboardingAnswers): String = when (id) {
    "sought_support" -> answers.soughtSupport
    "openness"       -> answers.openness
    else             -> ""
}

@HiltViewModel
class OnboardingViewModel @Inject constructor(
    private val onboardingRepo: OnboardingRepository,
) : ViewModel() {

    private val _state = MutableStateFlow(OnboardingUiState())
    val state: StateFlow<OnboardingUiState> = _state.asStateFlow()

    // ── Navigation ────────────────────────────────────────────────────────

    fun advance() {
        val s = _state.value
        if (s.isLastStep) return
        _state.update { it.copy(currentIndex = it.currentIndex + 1) }
    }

    fun back() {
        val s = _state.value
        if (s.isFirstStep) return
        _state.update { it.copy(currentIndex = it.currentIndex - 1) }
    }

    // ── Answer setters ────────────────────────────────────────────────────

    fun setUsername(value: String) =
        _state.update { it.copy(answers = it.answers.copy(username = value)) }

    fun setFeelingSentence(value: String) =
        _state.update { it.copy(answers = it.answers.copy(feelingSentence = value)) }

    fun toggleChipSelection(stepId: String, option: String, maxSelections: Int) {
        _state.update { s ->
            val answers = s.answers
            val updated = when (stepId) {
                "mood_descriptor" -> answers.copy(moodDescriptors = toggle(answers.moodDescriptors, option, maxSelections))
                "feeling_now"     -> answers.copy(feelingNow     = toggle(answers.feelingNow,     option, maxSelections))
                "feeling_like"    -> answers.copy(feelingLike    = toggle(answers.feelingLike,    option, maxSelections))
                "struggling_with" -> answers.copy(strugglingWith = toggle(answers.strugglingWith, option, maxSelections))
                "vibes"           -> answers.copy(vibes          = toggle(answers.vibes,          option, maxSelections))
                "hope_to_gain"    -> answers.copy(hopeToGain     = toggle(answers.hopeToGain,     option, maxSelections))
                else              -> answers
            }
            s.copy(answers = updated)
        }
    }

    fun setMoodColor(key: MoodColorKey) {
        val label = DefaultMoodPalette.firstOrNull { it.key == key }?.label ?: ""
        _state.update { it.copy(answers = it.answers.copy(moodColor = key, moodLabel = label)) }
    }

    fun setRadioAnswer(stepId: String, option: String) {
        _state.update { s ->
            val updated = when (stepId) {
                "sought_support" -> s.answers.copy(soughtSupport = option)
                "openness"       -> s.answers.copy(openness = option)
                else             -> s.answers
            }
            s.copy(answers = updated)
        }
    }

    // ── Completion ────────────────────────────────────────────────────────

    fun completeOnboarding(uid: String) {
        _state.update { it.copy(isSaving = true) }
        viewModelScope.launch {
            onboardingRepo.saveOnboardingAnswers(uid, _state.value.answers)
                .onSuccess { _state.update { it.copy(isComplete = true, isSaving = false) } }
                .onFailure { e -> _state.update { it.copy(error = e.message, isSaving = false) } }
        }
    }

    fun clearError() = _state.update { it.copy(error = null) }

    // ── Helpers ───────────────────────────────────────────────────────────

    private fun toggle(list: List<String>, item: String, max: Int): List<String> =
        if (item in list) list - item
        else if (max > 0 && list.size >= max) list
        else list + item
}
